### PR TITLE
feat: gateway can create normal invoices

### PR DESCRIPTION
**Pull request by tvolk131**

### CODE REVIEW

File: gateway/ln-gateway/src/bin/cln_extension.rs [hunk 1]
```diff
@@ -797,8 +847,14 @@ impl GatewayLightning for ClnRpcService {
             description,
         } = create_invoice_request.into_inner();
 
-        let payment_hash = sha256::Hash::from_slice(&payment_hash)
-            .map_err(|e| tonic::Status::internal(e.to_string()))?;
+        let payment_hash = if payment_hash.is_empty() {
+            return self
+                .create_invoice_with_known_preimage(amount_msat, expiry_secs.into(), description)
```
Comment (Kodylow) [inline gateway/ln-gateway/src/bin/cln_extension.rs:852]: Naming might be a little weird here: this means "I know the preimage so this is a payment to myself not on behalf of a fedimint client"

Comment (Kodylow) [reply gateway/ln-gateway/src/bin/cln_extension.rs:852]: for_self or something obviosuly "this is mine" might be better

File: gateway/ln-gateway/src/lib.rs [hunk 2]
```diff
@@ -920,6 +920,29 @@ impl Gateway {
...
+    ) -> Result<Bolt11Invoice> {
+        let GatewayState::Running { lightning_context } = self.get_state().await else {
+            return Err(GatewayError::Disconnected);
+        };
+
+        Bolt11Invoice::from_str(
+            &lightning_context
+                .lnrpc
+                .create_invoice(CreateInvoiceRequest {
+                    payment_hash: Vec::new(),
```
Comment (Kodylow) [inline gateway/ln-gateway/src/lib.rs:935]: add a comment here

**Review by Kodylow — APPROVED** [review]
lgtm, nits: change the naming scheme to have "self" to clarify when it's the gateway doing things for itself vs fedimint client actions

File: modules/fedimint-ln-common/src/gateway_endpoint_constants.rs [hunk 3]
```diff
@@ -17,6 +17,7 @@ pub const OPEN_CHANNEL_ENDPOINT: &str = "/open_channel";
 pub const CLOSE_CHANNELS_WITH_PEER_ENDPOINT: &str = "/close_channels_with_peer";
 pub const RECEIVE_ECASH_ENDPOINT: &str = "/receive_ecash";
 pub const ROUTING_INFO_V2_ENDPOINT: &str = "/routing_info";
+pub const CREATE_INVOICE_ENDPOINT: &str = "/create_invoice";
```
Comment (m1sterc001guy) [inline modules/fedimint-ln-common/src/gateway_endpoint_constants.rs:20]: Let's start migrating these to the lnv2 module

Comment (tvolk131) [reply modules/fedimint-ln-common/src/gateway_endpoint_constants.rs]: Good point! Done.

File: gateway/ln-gateway/src/lib.rs (hunk 2 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/lib.rs]: Done.

File: gateway/ln-gateway/src/bin/cln_extension.rs (hunk 1 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/bin/cln_extension.rs]: Done.

File: gateway/ln-gateway/src/lightning/ldk.rs [hunk 4]
```diff
@@ -363,33 +363,43 @@ impl ILnRpcClient for GatewayLdkClient {
...
+        // Currently `ldk-node` only supports direct descriptions.
+        // See https://github.com/lightningdevkit/ldk-node/issues/325.
+        // TODO: Once the above issue is resolved, we should support
+        // description hashes as well.
+        let description_str =
+            if let Some(Description::Direct(description)) = &create_invoice_request.description {
+                description
+            } else {
+                ""
+            };
```
Comment (elsirion) [inline gateway/ln-gateway/src/lightning/ldk.rs:385]: We should return an error on receiving a description hash. If someone does that they expect a specific behavior that we can't provide right now. It's better to fail.

File: gateway/ln-gateway/src/lib.rs [hunk 5]
```diff
@@ -920,6 +920,30 @@ impl Gateway {
...
+    ) -> Result<Bolt11Invoice> {
+        let GatewayState::Running { lightning_context } = self.get_state().await else {
+            return Err(GatewayError::Disconnected);
+        };
+
+        Bolt11Invoice::from_str(
+            &lightning_context
+                .lnrpc
+                .create_invoice(CreateInvoiceRequest {
+                    payment_hash: Vec::new(), /* Empty payment hash indicates an invoice payable
```
Comment (elsirion) [inline gateway/ln-gateway/src/lib.rs:935]: Can we make this more explicit? At least an `Option<Vec<u8>>`?

File: gateway/ln-gateway/src/bin/cln_extension.rs [hunk 6]
```diff
@@ -383,6 +383,56 @@ impl ClnRpcService {
...
+        expiry_secs: u64,
+        description_or: Option<Description>,
+    ) -> Result<tonic::Response<CreateInvoiceResponse>, Status> {
+        let description = match description_or {
+            None => String::new(),
+            Some(Description::Direct(description)) => description,
+            Some(Description::Hash(_hash)) => {
+                // Currently CLN only supports direct descriptions.
+                String::new()
+            }
```
Comment (elsirion) [inline gateway/ln-gateway/src/bin/cln_extension.rs:401]: Same here: better to fail

File: gateway/ln-gateway/src/lib.rs (hunk 5 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/lib.rs:935]: Nope 🙁 `CreateInvoiceRequest` is a proto message, so the [default value](https://protobuf.dev/programming-guides/proto3/#default) for bytes is an empty list of bytes.

File: gateway/ln-gateway/src/bin/cln_extension.rs (hunk 6 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/bin/cln_extension.rs]: Done.

File: gateway/ln-gateway/src/lightning/ldk.rs (hunk 4 above)
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs]: Done.

**Pull request merged**